lib/srv/db/cloud/watcher.go

```diff
@@ -0,0 +1,165 @@
+func (c *WatcherConfig) CheckAndSetDefaults() error {
+	if c.Clients == nil {
+		c.Clients = cloud.NewClients()
+	}
+	if c.Interval == 0 {
+		c.Interval = 5 * time.Minute
+	}
+	c.AzureMatchers = db.SimplifyMatchers(c.AzureMatchers)
+	return nil
+}
+
+// Watcher monitors cloud databases according to the provided selectors.
+type Watcher struct {
+	// cfg is the watcher config.
+	cfg WatcherConfig
+	// log is the watcher logger.
+	log logrus.FieldLogger
+	// ctx is the watcher close context.
+	ctx context.Context
+	// fetchers fetch databases according to their selectors.
+	fetchers []discovery.Fetcher
+	// databasesC is a channel where fetched databases are sent.
+	databasesC chan (types.Databases)
+}
+
+// NewWatcher returns a new instance of a cloud databases watcher.
+func NewWatcher(ctx context.Context, config WatcherConfig) (*Watcher, error) {
+	if err := config.CheckAndSetDefaults(); err != nil {
+		return nil, trace.Wrap(err)
+	}
+	fetchers, err := makeFetchers(ctx, &config)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	if len(fetchers) == 0 {
+		return nil, trace.NotFound("no cloud selectors")
+	}
+	return &Watcher{
+		cfg:        config,
+		log:        logrus.WithField(trace.Component, "watcher:cloud"),
+		ctx:        ctx,
+		fetchers:   fetchers,
+		databasesC: make(chan types.Databases),
+	}, nil
+}
+
+// Start starts fetching cloud databases and sending them to the channel.
+//
+// TODO(r0mant): In future, instead of (or in addition to) polling, we can
+// use a combination of EventBridge (former CloudWatch Events) and SQS/SNS to
+// subscribe to events such as created/removed instances and tag changes, but
+// this will require Teleport to have more AWS permissions.
+func (w *Watcher) Start() {
+	ticker := time.NewTicker(w.cfg.Interval)
+	defer ticker.Stop()
+	w.log.Debugf("Starting cloud databases watcher.")
+	w.fetchAndSend()
+	for {
+		select {
+		case <-ticker.C:
+			w.fetchAndSend()
+		case <-w.ctx.Done():
+			w.log.Debugf("Cloud databases watcher done.")
+			return
+		}
+	}
+}
+
+// fetchAndSend fetches databases from all fetchers and sends them to the channel.
+func (w *Watcher) fetchAndSend() {
+	var result types.Databases
+	for _, fetcher := range w.fetchers {
+		resources, err := fetcher.Get(w.ctx)
+		if err != nil {
+			// DB agent may have permissions to fetch some databases but not
+			// others. This is acceptable, thus continue to other fetchers.
+			// DB agent may also query for resources that do not exist. This is ok.
+			// If the resource is created in the future, we will fetch it then.
+			if trace.IsAccessDenied(err) || trace.IsNotFound(err) {
+				w.log.WithError(err).Debugf("Skipping fetcher %v.", fetcher)
+				continue
+			}
+
+			w.log.WithError(err).Errorf("%s failed.", fetcher)
+			return
+		}
+		databases, err := resources.AsDatabases()
+		if err != nil {
+			w.log.WithError(err).Errorf("Failed to convert types.ResourceWithLabels to types.Databases for fetcher %v.", fetcher)
+			return
+		}
```

Comment:
this file is moved from `lib/srv/cloud/watchers` to `lib/srv/cloud/`, and updated to use discovery.Fetchers.